lib/galaxy/datatypes/goldenpath.py

```diff
@@ -0,0 +1,597 @@
+"""
+MIT License
+
+Copyright (c) 2021 Michael Alonge <malonge11@gmail.com>
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import abc
+import os
+from typing import Set, Union
+
+
+class AGPError(Exception):
+    """ Exception raised for AGP related errors. """
+
+    def __init__(self, fname, line_number, message="Error in AGP file."):
+        self.fname = fname
+        self.line_number = line_number
+        self.message = message
+
+        self.report = f"\n\nFILE: {self.fname}\nLINE: {self.line_number}\nERROR: {self.message}"
+        super().__init__(self.report)
+
+    def __repr__(self):
+        return "AGPError"
+
+
+class AGPFile:
+
+    """
+    A class storing the contents of an AGP v2.1 file. https://www.ncbi.nlm.nih.gov/assembly/agp/AGP_Specification/
+
+    The class is able to read new AGP lines in order to sequentially build the complete file.
+
+    The class should be capable of checking the validity of the file, as well as writing the AGP contents
+    to a file stream.
+
+    Common abbreviations:
+      "comp": AGP component
+      "obj":  AGP object
+      "pid":  AGP part number
+    """
+
+    def __init__(self, in_file, mode="r"):
+        if mode != "r" and mode != "w":
+            raise ValueError("AGPFile mode must be either read ('r') or write ('w').")
+
+        self._agp_version = "2.1"
+        self._fname = os.path.abspath(in_file)
+
+        # Store comment and AGP lines separately.
+        self._comment_lines = []
+        self._objects = []
+
+        # Store info enabling us to keep track of the state of the AGP file
+        self._current_obj = None
+        self._seen_objs = set()
+
+        # Read the contents of the AGP file
+        if mode == "r":
+            self._read_file()
```

Comment:
Assuming we're trusting the liens as in your above comment, could we just read the first few lines?

Reply:
I guess so.

Reply:
I don't know, that's why I'm asking why we need a data provider, where the contents of the file are returned. If the data provider is going to be used it might be a good idea to just provide an iterator in of the file contents, instead of the full file contents in memory. In general it is a good idea to check incoming external things at the boundary with Galaxy, so here you could easily do the full check when setting metadata. Then you know that the file is valid and you don't need to check every time the data provider is used.